src/DependencyInjection/Compiler/FragmentRegistryPass.php

```diff
@@ -0,0 +1,148 @@
+    {
+        if (!$container->has(self::FRAGMENT_REGISTRY)) {
+            return;
+        }
+
+        $this->fragmentRegistry = $container->findDefinition(self::FRAGMENT_REGISTRY);
+
+        if (!$this->classImplementsInterface(
+            $this->fragmentRegistry->getClass(), FragmentRegistryInterface::class)
+        ) {
+            return;
+        }
+
+        foreach ([
+            self::TAG_FRAGMENT_FRONTEND_MODULE,
+            self::TAG_FRAGMENT_PAGE_TYPE,
+            self::TAG_FRAGMENT_CONTENT_ELEMENT,
+        ] as $tag) {
+            $this->registerFragments($container, $tag);
+        }
+
+        foreach ([
+            self::RENDERER_FRONTEND_MODULE => self::TAG_RENDERER_FRONTEND_MODULE,
+            self::RENDERER_PAGE_TYPE => self::TAG_RENDERER_PAGE_TYPE,
+            self::RENDERER_CONTENT_ELEMENT => self::TAG_FRAGMENT_CONTENT_ELEMENT,
+        ] as $renderer => $tag) {
+            $this->registerFragmentRenderers($container, $renderer, $tag);
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param string           $tag
+     */
+    private function registerFragments(ContainerBuilder $container, string $tag)
+    {
+        $fragments = $this->findAndSortTaggedServices($tag, $container);
+
+        foreach ($fragments as $priority => $reference) {
+            $fragment = $container->findDefinition($reference);
+            $fragmentOptions = $fragment->getTag($tag)[0];
+
+            if (!isset($fragmentOptions['type'])) {
+                throw new RuntimeException(sprintf('A service tagged as "%s" must have a "type" attribute set.', $tag));
+            }
+
+            $fragmentOptions['controller'] = (string) $reference;
+            $fragmentOptions['tag'] = $tag;
+
+            // Support specific method on controller
+            if (isset($fragmentOptions['method'])) {
+                $fragmentOptions['controller'] .= ':'.$fragmentOptions['method'];
+                unset($fragmentOptions['method']);
+            }
+
+            // Mark all fragments as lazy so they are lazy loaded using
+            // the proxy manager (which is why we need to require it in the
+            // composer.json (otherwise the lazy definition will just be ignored)
+            $fragment->setLazy(true);
+
+            $fragmentIdentifier = $tag.'.'.$fragmentOptions['type'];
+            $this->fragmentRegistry->addMethodCall('addFragment', [$fragmentIdentifier, $reference, $fragmentOptions]);
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param string           $renderer
+     * @param string           $tag
+     */
+    private function registerFragmentRenderers(ContainerBuilder $container, string $renderer, string $tag)
+    {
+        if (!$container->has($renderer)) {
+            return;
+        }
+
+        $renderer = $container->findDefinition($renderer);
+        $renderer->setArgument(0, $this->findAndSortTaggedServices($tag, $container));
```

Comment:
Not necessary the right place to comment this, but just came to my mind while reading this part. Symfony has a lot of these cases of delegating "something". Whenever there is only one option/tagged service, it does not actually use the delegating loader but set the alias to the only service. So if there is only the DefaultRenderer, we would not need a delegating renderer at all. Example: https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L824-L832

Reply:
you will get the actual service and it does not have to loop?

Reply:
It just makes things way harder to understand without providing any advantage at all. It's a complete nonsense to introduce something like this to avoid a single loop of which you have thousands in a regular app. The security component of Symfony does a lot of that stuff and it's an absolute nightmare to understand their dependency injection configuration.